pkg/typesystem/resolver.go

```diff
@@ -0,0 +1,79 @@
+package typesystem
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/karlseguin/ccache/v3"
+	"github.com/oklog/ulid/v2"
+	"github.com/openfga/openfga/pkg/storage"
+)
+
+const (
+	typesystemCacheTTL = 168 * time.Hour // 7 days
+)
+
+// TypesystemResolverFunc is a function that implementations can implement to provide lookup and
+// resolution of a Typesystem.
+type TypesystemResolverFunc func(ctx context.Context, storeID, modelID string) (*TypeSystem, error)
+
+// MemoizedTypesystemResolver returns a TypesystemResolverFunc that either fetches the provided authorization
+// model (if provided) or looks up the latest authorizaiton model, and then it constructs a TypeSystem from
+// the resolved model. The type-system resolution is memoized so if another lookup of the same model occurs,
+// then the earlier TypeSystem that was constructed will be used.
+//
+// The memoized resolver function is safe for concurrent use.
+func MemoizedTypesystemResolverFunc(reader storage.AuthorizationModelReadBackend) TypesystemResolverFunc {
```

Comment:
nitpit: weird naming convention 🤔

Reply:
How so? What would be your suggested name?

Reply:
You normally use `ds` or `datastore`, kinda got used to that 😆

Reply:
Oh, I see. Happy to change that. The change was because in most other places we are referring to `OpenFGADatastore` (e.g. the whole interface), whereas here we are just referring to the read interface for Authz models.

Reply:
Also please fix the doc: `MemoizedTypesystemResolver` to `MemoizedTypesystemResolverFunc` :)

The reason i bring it up is because Goland is underlining this for me and it's veeeery annoying
<img width="1038" alt="image" src="https://github.com/openfga/openfga/assets/5374887/4ff1c576-4cc7-4991-a632-712d4d0058af">

Reply:
Done and done 👍